### ISSUE TITLE
Failed to use mTLS with PHP Session handler (redis mode)

### ISSUE THREAD

**Issue by nusphere** [phpredis/phpredis#2062]
### I'm seeing this behaviour on
- OS: Ubuntu
- Redis: 6
- PHP: 7.4
- phpredis: 5.3

### Steps to reproduce, backtrace or example script

When setup a local redis server (via Docker) with mTLS it is not possible to use the ssl stream setup for the php ini setup. I tested this behavior via PHPUnit. 

```php

    /**
     * This test is working with the mTLS setup
     */
    public function testConnectionWithTLS(): void
    {
        $redis = new Redis();

        $cert = __DIR__ . DIRECTORY_SEPARATOR . 'client-cert.pem';
        $key = __DIR__ . DIRECTORY_SEPARATOR. 'client-key.pem';
        $ca = __DIR__ . DIRECTORY_SEPARATOR. 'ca.pem';

        $redis->connect('127.0.1.1', '6379', 0.0, null, 0, 0.0, [
            'stream' => [
                'local_cert' => $cert,
                'local_pk' => $key,
                'cafile' => $ca,
                'verify_peer_name' => false,
            ]
        ]);

        self::assertTrue($redis->ping());
    }
```

When testing the Session Handler, it seems there is no way to add stream settings for ssl/tls. i try to use https://github.com/phpredis/phpredis/issues/1782#issuecomment-688642159 - but this was only implemented for redis cluster. 

it should be easy to implent into the normal redis version.

```php
    /**
     * @runInSeparateProcess
     * @return void
     */
    public function testSessionHandlerWithTLSPath(): void
    {
        $cert = __DIR__ . DIRECTORY_SEPARATOR . 'client-cert.pem';
        $key = __DIR__ . DIRECTORY_SEPARATOR. 'client-key.pem';
        $ca = __DIR__ . DIRECTORY_SEPARATOR. 'ca.pem';

        $savePath = "tls://127.0.1.1:6379".
            "&stream[verify_peer_name]=0".
            "&stream[local_cert]=file://{$cert}".
            "&stream[local_pk]=file://{$key}".
            "&stream[cafile]=file://{$ca}";

        ini_set('session.save_handler', 'redis');
        ini_set('session.save_path', $savePath);

        session_start();
    }
```

![image](https://user-images.githubusercontent.com/5822825/150366737-877e8679-a018-4ced-81ed-4f8d720b2ba8.png)

### I've checked
- [X] There is no similar issue from other users
- [X] Issue isn't fixed in `develop` branch

**Comment by nusphere** [comment]
The only Workaround is to build a own Redis Session Handler:

```php
<?php

class RedisSessionHandler implements SessionHandlerInterface
{
    /**
     * @var Redis
     */
    private $redis;

    public function close()
    {
        return $this->redis->close();
    }

    public function destroy($id)
    {
        return $this->redis->del($id) === 1;
    }

    public function gc($max_lifetime)
    {
        // TODO: Implement gc() method.
    }

    public function open($path, $name)
    {
        $cert = __DIR__ . DIRECTORY_SEPARATOR . 'client-cert.pem';
        $key = __DIR__ . DIRECTORY_SEPARATOR. 'client-key.pem';
        $ca = __DIR__ . DIRECTORY_SEPARATOR. 'ca.pem';

        $this->redis = new Redis();
        $this->redis->connect('127.0.1.1', '6379', 0.0, null, 0, 0.0, [
            'stream' => [
                'local_cert' => $cert,
                'local_pk' => $key,
                'cafile' => $ca,
                'verify_peer_name' => false,
            ]
        ]);

        return $this->redis->ping();
    }

    public function read($id)
    {
        $result = $this->redis->get($id);

        return $result === false ? '' : $result;
    }

    public function write($id, $data)
    {
        $this->redis->set($id, $data);

        return $this->read($id) === $data;
    }
}
```

But this should not the solution, the existing https://github.com/phpredis/phpredis/blob/develop/redis_session.c implement a way more better behavior and should be used.

**Comment by nusphere** [comment]
https://github.com/phpredis/phpredis/blob/92540ad348ba4052dfc6ecf640a67c6459739c35/redis_session.c#L443-L451

should aware of the **stream** params and call `redis_sock_set_stream_context()` like for _rediscluster_

https://github.com/phpredis/phpredis/blob/92540ad348ba4052dfc6ecf640a67c6459739c35/redis_session.c#L943-L945

**Comment by LordRobinCbz** [comment]
Hello,

Does this ticket is in working state ? Enable mTLS for Redis "Standalone" / Sentinel replication cluster is very important for global security.

I didn't find a documentation about setting the mtls chain inside a php.ini file (php fpm) as an abstraction for php developers or applications server.

Any informations about this topic ?

Many thanks and best regards

**Comment by rhclayto** [comment]
Is this true still that you can't set tls options in `session.save_path` with standalone Redis? If so, please fix it.

Edit: It works. The correct string format was pointed out to me over at https://github.com/phpredis/phpredis/issues/2133#issuecomment-2773446123

**Comment by lcavana** [comment]
> Is this true still that you can't set tls options in `session.save_path` with standalone Redis? If so, please fix it.
> 
> Edit: It works. The correct string format was pointed out to me over at [#2133 (comment)](https://github.com/phpredis/phpredis/issues/2133#issuecomment-2773446123)

Absolutely not. This issue is for Mutual TLS in Redis Standalone connection strings; you were only trying to do server-side TLS; this clearly does not apply to you.
And by the way in phpredis 6.3.0 Mutual TLS **still** does not work.

**Comment by rhclayto** [comment]
> > Is this true still that you can't set tls options in `session.save_path` with standalone Redis? If so, please fix it.
> > Edit: It works. The correct string format was pointed out to me over at [#2133 (comment)](https://github.com/phpredis/phpredis/issues/2133#issuecomment-2773446123)
> 
> Absolutely not. This issue is for Mutual TLS in Redis Standalone connection strings; you were only trying to do server-side TLS; this clearly does not apply to you. And by the way in phpredis 6.3.0 Mutual TLS **still** does not work.

This issue is for mTLS in Redis Standalone connection strings for the phpredis ***session handler***. 
The OP's first test shows 'This test is working with the mTLS setup', but the second test showed it was not working with the session handler specifically. 

True, I haven't tried mTLS with the phpredis session handler. But if you read the text of the original post, nusphere states 'When testing the Session Handler, it seems there is no way to add stream settings for ssl/tls.' Which implies to me that the reason mTLS wasn't working was that it was impossible to add tls stream settings to the connection string for the session handler at all, which underlying problem is what brought me to this issue. This is no longer the case, fixed in the issue to which I linked.

Are you absolutely certain that the session handler connection string tls context options for mTLS (local_cert, local_pk) don't work, while all the other ones do?

**Comment by lcavana** [comment]
> This issue is for mTLS in Redis Standalone connection strings for the phpredis _**session handler**_. The OP's first test shows 'This test is working with the mTLS setup', but the second test showed it was not working with the session handler specifically.

Yes, it's for the session handler only. It's in the issue title so I took it for granted we are all speaking about this.

> True, I haven't tried mTLS with the phpredis session handler.

Then try, it won't work. It seems it's only implemented for rediscluster. And possibly the reason why this is issue is still open.

> This is no longer the case, fixed in the issue to which I linked.

It's only in the reproducer text as an example (besides, in the reproducer there are both mTLS and user/pass which I find odd, you'll typically want to use either one or the other, not both.)
